freqtrade/rpc/emc.py

```diff
@@ -0,0 +1,234 @@
+        """
+        Main connection loop for the consumer
+        """
+        try:
+            while True:
+                try:
+                    url, token = producer['url'], producer['ws_token']
+                    ws_url = f"{url}?token={token}"
+
+                    async with websockets.connect(ws_url) as ws:
+                        logger.info("Connection successful")
+                        channel = WebSocketChannel(ws)
+
+                        # Tell the producer we only want these topics
+                        # Should always be the first thing we send
+                        await channel.send(
+                            self.compose_consumer_request(RPCRequestType.SUBSCRIBE, self.topics)
+                        )
+
+                        # Now receive data, if none is within the time limit, ping
+                        while True:
+                            try:
+                                message = await asyncio.wait_for(
+                                    channel.recv(),
+                                    timeout=5
+                                )
+
+                                async with lock:
+                                    # Handle the data here
+                                    # We use a lock because it will call RPC methods
+                                    self.handle_producer_message(message)
+
+                            except (asyncio.TimeoutError, websockets.exceptions.ConnectionClosed):
+                                # We haven't received data yet. Check the connection and continue.
+                                try:
+                                    # ping
+                                    ping = await channel.ping()
+
+                                    await asyncio.wait_for(ping, timeout=self.ping_timeout)
+                                    logger.debug(f"Connection to {url} still alive...")
+
+                                    continue
+                                except Exception:
+                                    logger.info(
+                                        f"Ping error {url} - retrying in {self.sleep_time}s")
+                                    await asyncio.sleep(self.sleep_time)
+
+                                    break
+                            except Exception as e:
+                                logger.exception(e)
+                                continue
+                except (
+                    socket.gaierror,
+                    ConnectionRefusedError,
+                    websockets.exceptions.InvalidStatusCode
+                ) as e:
+                    logger.error(f"Connection Refused - {e} retrying in {self.sleep_time}s")
+                    await asyncio.sleep(self.sleep_time)
+
+                    continue
+
+        except asyncio.CancelledError:
+            # Exit silently
+            pass
+
+    def compose_consumer_request(self, type_: str, data: Any) -> Dict[str, Any]:
```

Comment:
should be `type_: RPCRequestType` - that'll make sure we'll not send random strings around

Reply:
:+1: